packages/abi-utils/lib/arbitrary.ts

```diff
@@ -0,0 +1,315 @@
+      fc
+        .tuple(
+          fc.oneof(
+            fc.constant("pure"),
+            fc.constant("view"),
+            fc.constant("nonpayable"),
+            fc.constant("payable")
+          ),
+          fc.boolean()
+        )
+        .map(([stateMutability, includeOptionals]) =>
+          !includeOptionals
+            ? { stateMutability }
+            : {
+                stateMutability,
+                payable: stateMutability === "payable",
+                constant:
+                  stateMutability === "view" || stateMutability === "pure"
+              }
+        )
+    )
+    .map(records => records.reduce((a, b) => ({ ...a, ...b }), {}))
+    .filter(entry => {
+      const { inputs, outputs = [] } = entry as Types.FunctionEntry;
+      // names that are not blank should be unique
+      const names = [...inputs, ...outputs]
+        .map(({ name }) => name)
+        .filter(name => name !== "");
+      return names.length === new Set(names).size;
+    });
+
+export const ReceiveEntry = () =>
+  fc.record({
+    type: fc.constant("receive"),
+    stateMutability: fc.constant("payable")
+  });
+
+export const FallbackEntry = () =>
+  fc
+    .tuple(
+      fc.record({
+        type: fc.constant("fallback")
+      }),
+      fc
+        .tuple(
+          fc.oneof(fc.constant("nonpayable"), fc.constant("payable")),
+          fc.boolean()
+        )
+        .map(([stateMutability, includeOptionals]) =>
+          !includeOptionals
+            ? { stateMutability }
+            : {
+                stateMutability,
+                payable: stateMutability === "payable",
+                constant:
+                  stateMutability === "view" || stateMutability === "pure"
+              }
+        )
+    )
+    .map(records => records.reduce((a, b) => ({ ...a, ...b }), {}));
+
+export const ConstructorEntry = () =>
+  fc
+    .tuple(
+      fc.record({
+        type: fc.constant("constructor"),
+        inputs: fc.array(Parameter())
+      }),
+      fc
+        .tuple(
+          fc.oneof(fc.constant("nonpayable"), fc.constant("payable")),
+          fc.boolean()
+        )
+        .map(([stateMutability, includeOptionals]) =>
+          !includeOptionals
+            ? { stateMutability }
+            : {
+                stateMutability,
+                payable: stateMutability === "payable",
+                constant:
+                  stateMutability === "view" || stateMutability === "pure"
+              }
+        )
+    )
+    .map(records => records.reduce((a, b) => ({ ...a, ...b }), {}));
+
+export const Abi = () =>
+  fc
+    .tuple(
+      ConstructorEntry(),
+      FallbackEntry(),
+      ReceiveEntry(),
+      fc.array(fc.oneof(FunctionEntry(), EventEntry()))
+    )
+    .chain(([constructor, fallback, receive, entries]) =>
+      fc.shuffledSubarray([constructor, fallback, receive, ...entries])
+    );
+
+namespace Numerics {
+  // 0 < n <= 32
+  export const Bytes = () => fc.nat(31).map(k => 32 - k);
```

Comment:
I think the `+1` trick that you use below would be a clearer expression of this than this subtraction trick.  Then it'd just be `fc.nat(31).map(k => k + 1)`.

Reply:
The reason I didn't do this is because fast-check maintains the concept of what is a simple value vs. what is a complex value (that's what that "shrink" business relates to). So I did `32 - k` because I want 32 to be the "simple" value (since 0 is the "simple" value from `fc.nat(31)`)

Reply:
Oh, huh.  In that case is a similar subtraction warranted for the decimal places?  80 is the default there, after all, not 1.

Reply:
Also this likely warrants a comment.

Reply:
Oops you are correct, I misremembered, sorry.  18 is the default.  They sound too similar in English I guess. :P  Um, IDK, do what you want with this I guess?